src/common/qat.c

```diff
@@ -0,0 +1,439 @@
+		D_ERROR("QAT: Failed to start user process SSL\n");
+		qaeMemDestroy();
+		return -1;
+	}
+
+	getDcInstance(dcInstHandle);
+	if (*dcInstHandle == NULL) {
+		D_ERROR("QAT: No DC instance\n");
+		return -1;
+	}
+
+	status = cpaDcQueryCapabilities(*dcInstHandle, &cap);
+	if (status != CPA_STATUS_SUCCESS)
+		return -1;
+
+	if (!cap.statelessDeflateCompression ||
+	    !cap.statelessDeflateDecompression ||
+	    !cap.checksumAdler32 ||
+	    !cap.dynamicHuffman)
+		return -1;
+
+	if (cap.dynamicHuffmanBufferReq) {
+		status = cpaDcBufferListGetMetaSize(*dcInstHandle, 1,
+						    &buffMetaSize);
+
+		if (status == CPA_STATUS_SUCCESS)
+			status = cpaDcGetNumIntermediateBuffers(
+				*dcInstHandle,
+				numInterBuffLists);
+
+		if (status == CPA_STATUS_SUCCESS && *numInterBuffLists != 0)
+			status = Mem_Alloc_Contig(
+				(void *)&interBufs,
+				*numInterBuffLists * sizeof(CpaBufferList *),
+				1);
+
+		/** if max buffer size is zero, set default value 64KB */
+		if (maxBufferSize == 0)
+			maxBufferSize = MAX_BUF_SIZE;
+
+		for (i = 0; i < *numInterBuffLists; i++) {
+			if (status == CPA_STATUS_SUCCESS)
+				status = Mem_Alloc_Contig(
+					(void *)&interBufs[i],
+					sizeof(CpaBufferList), 1);
+
+			if (status == CPA_STATUS_SUCCESS)
+				status = Mem_Alloc_Contig(
+				(void *)(&interBufs[i]->pPrivateMetaData),
+				buffMetaSize, 1);
+
+			if (status == CPA_STATUS_SUCCESS)
+				status = Mem_Alloc_Contig(
+					(void *)&interBufs[i]->pBuffers,
+					sizeof(CpaFlatBuffer), 1);
+
+			if (status == CPA_STATUS_SUCCESS) {
+				/* Implementation requires an intermediate */
+				/* buffer approximately twice the size of */
+				/* the output buffer */
+				status = Mem_Alloc_Contig(
+					(void *)&interBufs[i]->pBuffers->pData,
+					2 * maxBufferSize, 1);
+				interBufs[i]->numBuffers = 1;
+				interBufs[i]->pBuffers->dataLenInBytes =
+					2 * maxBufferSize;
+			}
+		}
+	}
+
+	if (status == CPA_STATUS_SUCCESS)
+		status = cpaDcSetAddressTranslation(*dcInstHandle, virtToPhys);
+
+	if (status == CPA_STATUS_SUCCESS)
+		status = cpaDcStartInstance(
+			*dcInstHandle, *numInterBuffLists, interBufs);
+
+	if (status == CPA_STATUS_SUCCESS) {
+		sd.compLevel = compLvl;
+		sd.compType = CPA_DC_DEFLATE;
+		sd.huffType = CPA_DC_HT_FULL_DYNAMIC;
+		sd.autoSelectBestHuffmanTree = CPA_DC_ASB_STATIC_DYNAMIC;
+		sd.sessDirection = CPA_DC_DIR_COMBINED;
+		sd.sessState = CPA_DC_STATELESS;
+		sd.checksum = CPA_DC_ADLER32;
+
+		status = cpaDcGetSessionSize(*dcInstHandle, &sd,
+					     &sess_size, &ctx_size);
+	}
+
+	if (status == CPA_STATUS_SUCCESS)
+		status = Mem_Alloc_Contig((void *)sessionHdl, sess_size, 1);
+
+	/* Initialize the Stateless session */
+	if (status == CPA_STATUS_SUCCESS) {
+		status = cpaDcInitSession(
+			*dcInstHandle,
+			*sessionHdl,
+			&sd,
+			NULL,
+			dcCallback);
+	}
+
+	*bufferInterArrayPtr = interBufs;
+
+	if (status == CPA_STATUS_SUCCESS)
+		return 0;
+
+	return -1;
+}
+
+int qat_dc_compress(CpaInstanceHandle *dcInstHandle,
```

Comment:
Maybe add some error logging so if qa_dc_compress returns 0 the logs would indicate what the failure was. Actually I'm wondering if daos_compressor_compress/decompress should be changed to return a negative on error? For example how would a caller know if the system's run out of error (Mem_Alloc_Contig fails) or if dcResults.produced > dstLen and compress just needs to be called again with a larger buffer?

Reply:
Added the return code.